### PR TITLE
Add AI Rubrics Experiment and Account Setting

### PR DESCRIPTION
**Pull request by wilkie**
This adds the AI Settings section and a boolean value to turn off the AI rubrics assessment. It is a "disable" setting since the value will be assumed at first to be on. (opt-out)

This is only available to verified teachers.

Uses the `ai-rubrics` experiment value that can be used to obscure the rubrics work unless a verified teacher exists within the pilot of that name.

You can block out such work via an experiment tag (here in some HAML):

```
- if Experiment.enabled?(user: current_user, experiment_name: 'ai-rubrics')
  %p
    This is only seen when the current user is in the ai-rubrics pilot
```

## Links

- jira ticket: [AITT-82](https://codedotorg.atlassian.net/browse/AITT-82)

## Testing story

Visually:

image

## Deployment strategy

An experiment tag that can be fully turned on via the `DCDO` set experiment functionality. For the pilot, adding verified teachers to the `ai-rubrics` pilot.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/views/devise/registrations/edit.html.haml [hunk 1]
```diff
@@ -115,6 +115,18 @@
       )
     %p= t('user.only_one_parent_email_supported_warning')
 
+- if experiment_value('ai_rubrics', request)
```
Comment (bethanyaconnor) [inline dashboard/app/views/devise/registrations/edit.html.haml]: ha, I named this experiment `ai-rubrics`. We should probably standardize. Happy to use `ai_rubrics` as I'll need to update my PR anyway

File: dashboard/app/views/devise/registrations/edit.html.haml [hunk 2]
```diff
@@ -115,6 +115,18 @@
...
+  - if current_user.verified_teacher?
```
Comment (bethanyaconnor) [inline dashboard/app/views/devise/registrations/edit.html.haml:135]: Sorry if I missed this but why only verified teachers?

**Comment by bethanyaconnor** [comment]
Sorry, I've never seen this experiment system before. Is the plan to mostly hide this behind a DCDO flag then release this setting to everyone at once? Or are we hoping to selectively show it only to pilot teachers?

If it's the latter, we might want to zoom out and settle on how we're doing our experiment flags.

File: dashboard/app/views/devise/registrations/edit.html.haml (hunk 2 above)
Comment (wilkie) [reply dashboard/app/views/devise/registrations/edit.html.haml:135]: It's not in the spec, so I can make it available to all teachers, true.

**Comment by wilkie** [comment]
Yes, it is a flag that turns it on for everybody. You can selectively turn it on via a query parameter `?ai_rubrics=1`, etc. Or a cookie where `ai_rubrics` is set. For a pilot, we can selectively turn it on for a teacher in the pilot by having a route that adds the ai_rubrics cookie.

**Comment by wilkie** [comment]
I'm following an example set by the Platform team that they used for an offline pilot and for the CPA and gender field work. I'm open to other ways.

File: dashboard/app/views/devise/registrations/edit.html.haml (hunk 2 above)
Comment (bethanyaconnor) [reply dashboard/app/views/devise/registrations/edit.html.haml:135]: I don't have a strong opinion here so whatever you think is best! Was just curious as to how we came to that decision

Comment (wilkie) [reply dashboard/app/views/devise/registrations/edit.html.haml:135]: We should revisit this question! I think we were again considering verified teachers only for now. @bethanyaconnor should we do this check here, too?

Comment (bethanyaconnor) [reply dashboard/app/views/devise/registrations/edit.html.haml:135]: I'll also tag @tess323 here but, yeah, I think it makes sense to check for verified teacher permissions here

File: dashboard/app/controllers/registrations_controller.rb [hunk 3]
```diff
@@ -445,6 +445,7 @@ def existing_account
       :provider,
       :us_state,
       :country_code,
+      :ai_rubrics_disabled,
```
Comment (bethanyaconnor) [inline dashboard/app/controllers/registrations_controller.rb:473]: Should AI features be opt in or opt out? Full disclosure, I think they should be opt-in but I'd want @tess323 to make the final decision there

**Comment by bethanyaconnor** [comment]
I checked this out locally and I can see the new setting when I add `?ai-rubrics=1` to the URL but I can't seem to get it to show up without it. I tried using the client-side experiment flag (I know that isn't supported) and adding the user to the pilot. Is that expected?

We will have another way to update this field, so this might be okay for the pilot.

File: dashboard/app/controllers/registrations_controller.rb (hunk 3 above)
Comment (wilkie) [reply dashboard/app/controllers/registrations_controller.rb:473]: I would like it to be opt in, but it felt like it was opt out from the descriptions. Will the AI popup opt people in?

**Comment by wilkie** [comment]
Let me test it, too. It would be enabled via the DCDO tag, I thought. But for the pilot, it would make sense to special case the participants? I'll have to ask you how to do that. :) Should be an easy fix.

**Comment by bethanyaconnor** [comment]
I didn't test with DCDO but I'd believe that would work! I'm wondering if it would work to use `Experiment.enabled?(user: current_user, experiment_name: 'ai-rubrics')` so that folks in the pilot can see it before we flip the DCDO flag?

To test the experiment locally, sign in as an admin, go to http://localhost-studio.code.org:3000/admin/pilots, create a pilot with the name `ai-rubrics`, then add a test account to that pilot.

File: dashboard/app/controllers/registrations_controller.rb (hunk 3 above)
Comment (bethanyaconnor) [reply dashboard/app/controllers/registrations_controller.rb:473]: The rubric in the level could be used for both opt-in and opt-out I think. By "descriptions", do you mean a product spec?

Comment (bethanyaconnor) [reply dashboard/app/controllers/registrations_controller.rb:473]: Acknowleding that we got an answer in Slack to be opt-out here

File: dashboard/test/ui/features/step_definitions/account_steps.rb [hunk 4]
```diff
@@ -116,6 +116,25 @@ def create_user(name, url: '/users.json', code: 201, **user_opts)
   end
 end
 
+def admin_user
```
Comment (bethanyaconnor) [inline dashboard/test/ui/features/step_definitions/account_steps.rb]: 🎉

**Review by bethanyaconnor — APPROVED** [review]
Thanks for adding tests! Looks great!

File: dashboard/test/ui/features/step_definitions/account_steps.rb (hunk 4 above)
Comment (wilkie) [reply dashboard/test/ui/features/step_definitions/account_steps.rb]: Oh! Actually... is this reasonable? It adds a floating admin user to the database... I could randomize the password each time... but it is useful for having the pilot flow integrated into our UI (acceptance) tests.

**Comment by wilkie** [comment]
> Thanks for adding tests! Looks great!

Thank you @bethanyaconnor for seeing this and reviewing even though I pushed so late! I appreciate that. I know it's falling out of scope and it's all late already, but, can you give me an opinion on adding an admin user to UI tests? I think this would persist an admin user with a predictable password in the context the tests run in... and these tests never run on production, I think. But worth getting your thoughts.

**Comment by wilkie** [comment]
It is a moot point. Can't create an admin account in anything except `develop` environment. Going to have to force the pilot creation instead.

**Comment by wilkie** [comment]
I'm adding just methods to create pilots programmatically. Unfortunately, I've had to reseed my test database so I'm waiting on that. 🙃 🥲

**Comment by bethanyaconnor** [comment]
Sorry sorry -- I have been so distracted with all the in person stuff! I wonder if you could create an endpoint to instead create a pilot (only in the test environment) and have the pilot be joinable by URL? Then the fake teacher could easily join the pilot from UI test steps.

**Comment by wilkie** [comment]
That's an interesting idea.

**Comment by wilkie** [comment]
That was a great point in the right direction! Thank you so much for that. There's an existing pattern of avoiding low-level work in the UI tests by having a very flexible test api in `TestController`. I used that to help create pilots in UI tests. Then, I can use the normal URL to add a user to the pilot. Seems to work and I would think the CI run would also pass now.

**Comment by wilkie** [comment]
CI failed with a unit test. I haven't added a unit test, so I assume it is flaky. It is this one:

```
= FAIL["test_throws_an_error_if_the_API_returns_a_non-200_response", "LtiAdvantageClientTest", 833.5877329579962]
--
1766 | test_throws_an_error_if_the_API_returns_a_non-200_response#LtiAdvantageClientTest (833.59s)
1767 | [RuntimeError] exception expected, not
1768 | Class: <ArgumentError>
1769 | Message: <"Private key required to sign JWT">
1770 | ---Backtrace---
1771 | lib/lti_access_token.rb:53:in `sign_jwt'
1772 | lib/lti_access_token.rb:29:in `get_access_token'
1773 | lib/clients/lti_advantage_client.rb:18:in `get_context_membership'
1774 | test/lib/clients/lti_advantage_client_test.rb:11:in `block (2 levels) in <class:LtiAdvantageClientTest>'
1775 | test/lib/clients/lti_advantage_client_test.rb:10:in `block in <class:LtiAdvantageClientTest>'
1776 | test/testing/setup_all_and_teardown_all.rb:36:in `run'
1777 | ---------------
1778 | test/lib/clients/lti_advantage_client_test.rb:10:in `block in <class:LtiAdvantageClientTest>'
1779 | test/testing/setup_all_and_teardown_all.rb:36:in `run'
```

**Comment by wilkie** [comment]
Rebased.

**Pull request merged**